zapcore/sampler.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap/internal/hash"
+)
+
+const (
+	_numLevels        = maxLevel - minLevel
+	_countersPerLevel = 4096
+)
+
+func newCounters() *counters {
+	return &counters{}
+}
+
+type counters [_numLevels][_countersPerLevel]uint64
```

Comment:
It's not a struct with a pointer field, the field is a value.
https://github.com/uber-go/atomic/blob/master/atomic.go#L175

The memory layout will be exactly the same, there should be no indirection either way.

The only performance difference will be because of the extra function call, which Go should be inlining but doesn't yet ([functions that call other functions are not inlined](https://groups.google.com/forum/#!topic/golang-nuts/V_xI29FGDZM)). Unfortunately that will probably add a handful of nanoseconds.

We should probably have benchmarks in the atomic package itself to measure the overhead.

The atomic package only really provides a couple of things:
 - safety: since you can't access the underlying value directly, you have to use the atomic accessors. This avoids accidentally doing a non-atomic read of an atomic value, which causes a data race.
 - readability: rather than `atomic.AddUint64(&c[i][j], 1)`, the code becomes `c[i][j].Inc()`